### DISCUSSION TITLE
Automatic cancelation of derived units

### DISCUSSION THREAD

**Discussion by dkirkby** [josdejong/mathjs#2658] [Q&A] [answered]
When I multiply two quantities, I would expect that any  unambiguous cancellations of derived units would be applied automatically, however this example using mathjs@7.2.0 returns `"(J m) / (m s)"` instead of the expected `"J / s"`:
```
{
  let a=math.unit(1, 'J/m');
  let b=math.unit(1, 'm/s');
  return math.multiply(a,b).toJSON().unit;
}
```
If there are performance reasons to defer this type of automatic simplification, perhaps a method to manually trigger would be useful  (and apologies if it already exists and I missed it).  The `toSI`  method is not what I am looking for  since it returns `"(kg m^2) / s^3"` which is less useful than `"J / s"` in many contexts.  

I  noticed the `skipAutomaticSimplification` attribute, but setting this `false` on `a` and `b` before multiplying does not change the result.

**Comment by josdejong** [comment]
In most cases the units will be simplified. @ericman314 do you have an idea why `toSI` returns `(kg m^2) / s^3` in this case?

**Comment by ericman314** [comment, marked as answer]
I'm not sure. If anything, I would think it should return `W` since that is even simpler than `J / s`. It doesn't really "know" that it should cancel the m's, it just looks at the entire unit to see if there is a single unit (like W) that fits.

Units are not actually simplified until they are formatted for output; this does improve the performance if you have lots of calculations in a row, and output one answer at the end. But this also means `toString` has side effects--ouch!

Because of these problems, I've been working on a complete overhaul of the units for a while now, and when that is ready it should address issues like this.

**Comment by josdejong** [comment]
Thanks for your feedback. Can we already try this example out in [UnitMath](https://github.com/ericman314/UnitMath)? :)

**Comment by ericman314** [comment]
Yes--I believe it should be working there.

**Comment by dkirkby** [comment]
> @ericman314 do you have an idea why toSI returns (kg m^2) / s^3 in this case?

That is the correct answer here since `toSI` should reduce to the unique combination of base SI units.  That's not always useful for human consumption, but needed as a canonical form, e.g. to test equality of different dimensioned expressions.
